src/impl_/wrap.rs

```diff
@@ -40,11 +41,44 @@ where
     T: IntoPy<PyObject>,
 {
     type Error = E;
-    fn wrap(self, py: Python<'_>) -> Result<Py<PyAny>, Self::Error> {
-        self.map(|o| o.into_py(py))
+    #[inline]
+    fn wrap(self) -> Result<T, Self::Error> {
+        self
     }
 }
 
+/// This is a follow-up function to `OkWrap::wrap` that converts the result into
+/// a `*mut ffi::PyObject` pointer.
+pub fn map_result_into_ptr<T: IntoPy<PyObject>>(
+    py: Python<'_>,
+    result: PyResult<T>,
+) -> PyResult<*mut ffi::PyObject> {
+    result.map(|obj| obj.into_py(py).into_ptr())
+}
+
+/// This is a follow-up function to `OkWrap::wrap` that converts the result into
+/// a safe wrapper.
+pub fn map_result_into_py<T: IntoPy<PyObject>>(
+    py: Python<'_>,
+    result: PyResult<T>,
+) -> PyResult<PyObject> {
+    result.map(|err| err.into_py(py))
+}
```

Comment:
```suggestion
pub fn map_result_into_ptr<T, E>(
    py: Python<'_>,
    result: Result<T, E>,
) -> PyResult<*mut ffi::PyObject>
where
    T: IntoPy<PyObject>,
    PyErr: From<E>,
{
    Ok(result?.into_py(py).into_ptr())
}

/// This is a follow-up function to `OkWrap::wrap` that converts the result into
/// a safe wrapper.
pub fn map_result_into_py<T, E>(
    py: Python<'_>,
    result: Result<T, E>,
) -> PyResult<PyObject>
where
    T: IntoPy<PyObject>,
    PyErr: From<E>,
{
    Ok(result?.into_py(py))
}
```

Reply:
I guess it's a bit of a style thing, I slightly prefer the combinators over `?` in these one-line implementations.

Reply:
The suggestion is not about the `?` but using `Result<T, E>` instead of `PyResult<T>`. If you use `map_err` instead of `?`, you can in fact relax the bound to `E: Into<PyErr>` instead of `PyErr: From<E>`, and that's maybe better (I should also relax the bound in async support implementation by the way)

Reply:
Ah I see. I take your point (and I did initially accept `Result<T, E>`) but having the `E: Into<PyErr>` bound introduced on this function and doing `.map_err` in here instead of directly in the macro generated code means this internal function `map_result_into_ptr` appears in the compiler error.

I didn't think that was useful (that was the UI test regression noted in the comment below) so I kept with just `PyResult` for now as all we need.

I nearly changed the async support but actually I think it's better if I don't touch that in this PR so that it conflicts less with what you've got stacked up.

Reply:
> [...] means this internal function `map_result_into_ptr` appears in the compiler error.

Makes sense indeed.